File: apps/apple/WorldInspector/Sources/WorldInspector/Protocol.swift
/// World Inspector Protocol — Swift types matching the Bevy WebSocket protocol.
///
/// These Codable types mirror the Rust `protocol.rs` definitions exactly,
/// enabling JSON serialization/deserialization over WebSocket.

import Foundation

// MARK: - Client → Server Messages

/// Messages sent from the inspector client to the Bevy server.
public enum ClientMessage: Encodable {
    case subscribe(topics: [String])
    case requestSceneTree
    case requestEntityDetail(entityId: UInt64)
    case requestWorldInfo
    case selectEntity(entityId: UInt64)
    case deselect
    case toggleVisibility(entityId: UInt64)
    case focusEntity(entityId: UInt64)
    case requestSceneSnapshot

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .subscribe(let topics):
            try container.encode("subscribe", forKey: .type)
            try container.encode(topics, forKey: .topics)
        case .requestSceneTree:
            try container.encode("request_scene_tree", forKey: .type)
        case .requestEntityDetail(let id):
            try container.encode("request_entity_detail", forKey: .type)
            try container.encode(id, forKey: .entityId)
        case .requestWorldInfo:
            try container.encode("request_world_info", forKey: .type)
        case .selectEntity(let id):
            try container.encode("select_entity", forKey: .type)
            try container.encode(id, forKey: .entityId)
        case .deselect:
            try container.encode("deselect", forKey: .type)
        case .toggleVisibility(let id):
            try container.encode("toggle_visibility", forKey: .type)
            try container.encode(id, forKey: .entityId)
        case .focusEntity(let id):
            try container.encode("focus_entity", forKey: .type)
            try container.encode(id, forKey: .entityId)
        case .requestSceneSnapshot:
            try container.encode("request_scene_snapshot", forKey: .type)
        }
    }

    private enum CodingKeys: String, CodingKey {
        case type
        case topics
        case entityId = "entity_id"
    }
}

// MARK: - Server → Client Messages

/// Messages received from the Bevy server.
public enum ServerMessage: Decodable {
    case sceneTree(entities: [TreeEntity])
    case entityDetail(entityId: UInt64, data: EntityDetailData)
    case worldInfo(data: WorldInfoData)
    case selectionChanged(entityId: UInt64)
    case selectionCleared
    case sceneChanged
    case entityTransformUpdated(entityId: UInt64, position: SIMD3<Float>, rotationDegrees: SIMD3<Float>)

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .type)

        switch type {
        case "scene_tree":
            let entities = try container.decode([TreeEntity].self, forKey: .entities)
            self = .sceneTree(entities: entities)
        case "entity_detail":
            let id = try container.decode(UInt64.self, forKey: .entityId)
            let data = try container.decode(EntityDetailData.self, forKey: .data)
            self = .entityDetail(entityId: id, data: data)
        case "world_info":
            let data = try container.decode(WorldInfoData.self, forKey: .data)
            self = .worldInfo(data: data)
        case "selection_changed":
            let id = try container.decode(UInt64.self, forKey: .entityId)
            self = .selectionChanged(entityId: id)
        case "selection_cleared":
            self = .selectionCleared
        case "scene_changed":
            self = .sceneChanged
        case "entity_transform_updated":
            let id = try container.decode(UInt64.self, forKey: .entityId)
            let pos = try container.decode([Float].self, forKey: .position)
            let rot = try container.decode([Float].self, forKey: .rotationDegrees)
            self = .entityTransformUpdated(
                entityId: id,
                position: SIMD3(pos[0], pos[1], pos[2]),
                rotationDegrees: SIMD3(rot[0], rot[1], rot[2])
            )
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .type, in: container,
                debugDescription: "Unknown message type: \(type)"
            )
        }
    }

    private enum CodingKeys: String, CodingKey {
        case type
        case entities
        case entityId = "entity_id"
        case data
        case position
        case rotationDegrees = "rotation_degrees"
    }
}

// MARK: - Data Types

public struct TreeEntity: Codable, Identifiable, Hashable {
    public let id: UInt64
    public let name: String
    public let entityType: String
    public let parentId: UInt64?
    public let visible: Bool
    public let children: [UInt64]

    enum CodingKeys: String, CodingKey {
        case id, name, visible, children
        case entityType = "entity_type"
        case parentId = "parent_id"
    }
}

public struct EntityDetailData: Codable {
    public let identity: IdentitySection
    public let transform: TransformSection?
    public let shape: String?
    public let material: MaterialSection?
    public let light: LightSection?
    public let behaviors: [BehaviorSection]
    public let audio: AudioSection?
    public let meshAsset: String?
    public let hierarchy: HierarchySection

    enum CodingKeys: String, CodingKey {
        case identity, transform, shape, material, light, behaviors, audio, hierarchy
        case meshAsset = "mesh_asset"
    }
}

public struct IdentitySection: Codable {
    public let name: String
    public let id: UInt64
    public let entityType: String

    enum CodingKeys: String, CodingKey {
        case name, id
        case entityType = "entity_type"
    }
}

public struct TransformSection: Codable {
    public let position: [Float]
    public let rotationDegrees: [Float]
    public let scale: [Float]
    public let visible: Bool

    enum CodingKeys: String, CodingKey {
        case position, scale, visible
        case rotationDegrees = "rotation_degrees"
    }
}

public struct MaterialSection: Codable {
    public let baseColor: [Float]
    public let metallic: Float
    public let roughness: Float
    public let reflectance: Float
    public let emissive: [Float]
    public let alphaMode: String
    public let doubleSided: Bool
    public let unlit: Bool

    enum CodingKeys: String, CodingKey {
        case metallic, roughness, reflectance, emissive, unlit
        case baseColor = "base_color"
        case alphaMode = "alpha_mode"
        case doubleSided = "double_sided"
    }
}

public struct LightSection: Codable {
    public let lightType: String
    public let color: [Float]
    public let intensity: Float
    public let range: Float?
    public let shadowsEnabled: Bool
    public let innerAngle: Float?
    public let outerAngle: Float?

    enum CodingKeys: String, CodingKey {
        case color, intensity, range
        case lightType = "light_type"
        case shadowsEnabled = "shadows_enabled"
        case innerAngle = "inner_angle"
        case outerAngle = "outer_angle"
    }
}

public struct BehaviorSection: Codable, Identifiable {
    public let id: String
    public let behaviorType: String
    public let basePosition: [Float]
    public let baseScale: [Float]

    enum CodingKeys: String, CodingKey {
        case id
        case behaviorType = "behavior_type"
        case basePosition = "base_position"
        case baseScale = "base_scale"
    }
}

public struct AudioSection: Codable {
    public let soundType: String
    public let volume: Float
    public let radius: Float
    public let attachedTo: String?
    public let position: [Float]?

    enum CodingKeys: String, CodingKey {
        case volume, radius, position
        case soundType = "sound_type"
        case attachedTo = "attached_to"
    }
}

public struct HierarchySection: Codable {
    public let parent: String?
    public let children: [String]
}

public struct WorldInfoData: Codable {
    public let name: String?
    public let entityCount: Int
    public let behaviorState: BehaviorStateInfo
    public let audio: AudioStateInfo?

    enum CodingKeys: String, CodingKey {
        case name, audio
        case entityCount = "entity_count"
        case behaviorState = "behavior_state"
    }
}

public struct BehaviorStateInfo: Codable {
    public let paused: Bool
    public let elapsed: Double
}

public struct AudioStateInfo: Codable {
    public let active: Bool
    public let emitterCount: Int
    public let ambienceLayers: [String]

    enum CodingKeys: String, CodingKey {
        case active
        case emitterCount = "emitter_count"
        case ambienceLayers = "ambience_layers"
    }
}

File: apps/android/world-inspector/src/main/java/com/localgpt/inspector/Protocol.kt
/// World Inspector Protocol — Kotlin types matching the Bevy WebSocket protocol.

package com.localgpt.inspector

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonContentPolymorphicSerializer
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

// ---------------------------------------------------------------------------
// JSON configuration
// ---------------------------------------------------------------------------

val InspectorJson = Json {
    ignoreUnknownKeys = true
    encodeDefaults = true
}

// ---------------------------------------------------------------------------
// Client → Server messages
// ---------------------------------------------------------------------------

@Serializable
sealed class ClientMessage {
    abstract val type: String

    @Serializable
    data class Subscribe(val topics: List<String>) : ClientMessage() {
        override val type = "subscribe"
    }

    @Serializable
    class RequestSceneTree : ClientMessage() {
        override val type = "request_scene_tree"
    }

    @Serializable
    data class RequestEntityDetail(
        @SerialName("entity_id") val entityId: ULong
    ) : ClientMessage() {
        override val type = "request_entity_detail"
    }

    @Serializable
    class RequestWorldInfo : ClientMessage() {
        override val type = "request_world_info"
    }

    @Serializable
    data class SelectEntity(
        @SerialName("entity_id") val entityId: ULong
    ) : ClientMessage() {
        override val type = "select_entity"
    }

    @Serializable
    class Deselect : ClientMessage() {
        override val type = "deselect"
    }

    @Serializable
    data class ToggleVisibility(
        @SerialName("entity_id") val entityId: ULong
    ) : ClientMessage() {
        override val type = "toggle_visibility"
    }

    @Serializable
    data class FocusEntity(
        @SerialName("entity_id") val entityId: ULong
    ) : ClientMessage() {
        override val type = "focus_entity"
    }

    @Serializable
    class RequestSceneSnapshot : ClientMessage() {
        override val type = "request_scene_snapshot"
    }
}

// ---------------------------------------------------------------------------
// Server → Client messages
// ---------------------------------------------------------------------------

@Serializable(with = ServerMessageSerializer::class)
sealed class ServerMessage {

    @Serializable
    data class SceneTree(val entities: List<TreeEntity>) : ServerMessage()

    @Serializable
    data class EntityDetail(
        @SerialName("entity_id") val entityId: ULong,
        val data: EntityDetailData
    ) : ServerMessage()

    @Serializable
    data class WorldInfo(val data: WorldInfoData) : ServerMessage()

    @Serializable
    data class SelectionChanged(
        @SerialName("entity_id") val entityId: ULong
    ) : ServerMessage()

    @Serializable
    object SelectionCleared : ServerMessage()

    @Serializable
    object SceneChanged : ServerMessage()

    @Serializable
    data class EntityTransformUpdated(
        @SerialName("entity_id") val entityId: ULong,
        val position: List<Float>,
        @SerialName("rotation_degrees") val rotationDegrees: List<Float>
    ) : ServerMessage()
}

object ServerMessageSerializer : JsonContentPolymorphicSerializer<ServerMessage>(ServerMessage::class) {
    override fun selectDeserializer(element: JsonElement) = when (
        element.jsonObject["type"]?.jsonPrimitive?.content
    ) {
        "scene_tree" -> ServerMessage.SceneTree.serializer()
        "entity_detail" -> ServerMessage.EntityDetail.serializer()
        "world_info" -> ServerMessage.WorldInfo.serializer()
        "selection_changed" -> ServerMessage.SelectionChanged.serializer()
        "selection_cleared" -> ServerMessage.SelectionCleared.serializer()
        "scene_changed" -> ServerMessage.SceneChanged.serializer()
        "entity_transform_updated" -> ServerMessage.EntityTransformUpdated.serializer()
        else -> throw IllegalArgumentException("Unknown type: ${element.jsonObject["type"]}")
    }
}

// ---------------------------------------------------------------------------
// Data types
// ---------------------------------------------------------------------------

@Serializable
data class TreeEntity(
    val id: ULong,
    val name: String,
    @SerialName("entity_type") val entityType: String,
    @SerialName("parent_id") val parentId: ULong? = null,
    val visible: Boolean,
    val children: List<ULong>
)

@Serializable
data class EntityDetailData(
    val identity: IdentitySection,
    val transform: TransformSection? = null,
    val shape: String? = null,
    val material: MaterialSection? = null,
    val light: LightSection? = null,
    val behaviors: List<BehaviorSection> = emptyList(),
    val audio: AudioSection? = null,
    @SerialName("mesh_asset") val meshAsset: String? = null,
    val hierarchy: HierarchySection
)

@Serializable
data class IdentitySection(
    val name: String,
    val id: ULong,
    @SerialName("entity_type") val entityType: String
)

@Serializable
data class TransformSection(
    val position: List<Float>,
    @SerialName("rotation_degrees") val rotationDegrees: List<Float>,
    val scale: List<Float>,
    val visible: Boolean
)

@Serializable
data class MaterialSection(
    @SerialName("base_color") val baseColor: List<Float>,
    val metallic: Float,
    val roughness: Float,
    val reflectance: Float,
    val emissive: List<Float>,
    @SerialName("alpha_mode") val alphaMode: String,
    @SerialName("double_sided") val doubleSided: Boolean,
    val unlit: Boolean
)

@Serializable
data class LightSection(
    @SerialName("light_type") val lightType: String,
    val color: List<Float>,
    val intensity: Float,
    val range: Float? = null,
    @SerialName("shadows_enabled") val shadowsEnabled: Boolean,
    @SerialName("inner_angle") val innerAngle: Float? = null,
    @SerialName("outer_angle") val outerAngle: Float? = null
)

@Serializable
data class BehaviorSection(
    val id: String,
    @SerialName("behavior_type") val behaviorType: String,
    @SerialName("base_position") val basePosition: List<Float>,
    @SerialName("base_scale") val baseScale: List<Float>
)

@Serializable
data class AudioSection(
    @SerialName("sound_type") val soundType: String,
    val volume: Float,
    val radius: Float,
    @SerialName("attached_to") val attachedTo: String? = null,
    val position: List<Float>? = null
)

@Serializable
data class HierarchySection(
    val parent: String? = null,
    val children: List<String> = emptyList()
)

@Serializable
data class WorldInfoData(
    val name: String? = null,
    @SerialName("entity_count") val entityCount: Int,
    @SerialName("behavior_state") val behaviorState: BehaviorStateInfo,
    val audio: AudioStateInfo? = null
)

@Serializable
data class BehaviorStateInfo(
    val paused: Boolean,
    val elapsed: Double
)

@Serializable
data class AudioStateInfo(
    val active: Boolean,
    @SerialName("emitter_count") val emitterCount: Int,
    @SerialName("ambience_layers") val ambienceLayers: List<String>
)
